alvr/sockets/src/stream_socket/mod.rs

```diff
@@ -0,0 +1,391 @@
+                port,
+                stream_tos_config,
+                send_buffer_bytes,
+                recv_buffer_bytes,
+            )?),
+            SocketProtocol::Tcp => StreamSocketBuilder::Tcp(tcp::bind(
+                timeout,
+                port,
+                stream_tos_config,
+                send_buffer_bytes,
+                recv_buffer_bytes,
+            )?),
+        })
+    }
+
+    pub fn accept_from_server(
+        self,
+        server_ip: IpAddr,
+        port: u16,
+        max_packet_size: usize,
+        timeout: Duration,
+    ) -> ConResult<StreamSocket> {
+        let (send_socket, receive_socket) = match self {
+            StreamSocketBuilder::Udp(socket) => {
+                udp::connect(&socket, server_ip, port, timeout).to_con()?;
+                udp::split_multiplexed(socket, max_packet_size).to_con()?
+            }
+            StreamSocketBuilder::Tcp(listener) => {
+                let socket = tcp::accept_from_server(&listener, Some(server_ip), timeout)?;
+                tcp::split_multiplexed(socket, timeout).to_con()?
+            }
+        };
+
+        Ok(StreamSocket {
+            send_socket: Arc::new(Mutex::new(send_socket)),
+            receive_socket,
+            queues: HashMap::new(),
+        })
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    pub fn connect_to_client(
+        timeout: Duration,
+        client_ip: IpAddr,
+        port: u16,
+        protocol: SocketProtocol,
+        dscp: Option<DscpTos>,
+        send_buffer_bytes: SocketBufferSize,
+        recv_buffer_bytes: SocketBufferSize,
+        max_packet_size: usize,
+    ) -> ConResult<StreamSocket> {
+        let (send_socket, receive_socket) = match protocol {
+            SocketProtocol::Udp => {
+                let socket =
+                    udp::bind(port, dscp, send_buffer_bytes, recv_buffer_bytes).to_con()?;
+                udp::connect(&socket, client_ip, port, timeout).to_con()?;
+                udp::split_multiplexed(socket, max_packet_size).to_con()?
+            }
+            SocketProtocol::Tcp => {
+                let socket = tcp::connect_to_client(
+                    timeout,
+                    &[client_ip],
+                    port,
+                    send_buffer_bytes,
+                    recv_buffer_bytes,
+                )?;
+                tcp::split_multiplexed(socket, timeout).to_con()?
+            }
+        };
+
+        Ok(StreamSocket {
+            send_socket: Arc::new(Mutex::new(send_socket)),
+            receive_socket,
+            queues: HashMap::new(),
+        })
+    }
+}
+
+pub struct StreamSocket {
+    send_socket: Arc<Mutex<Box<dyn MultiplexedSocketWriter + Send>>>,
+    receive_socket: Box<dyn MultiplexedSocketReader + Send>,
+    queues: HashMap<u16, StreamRecvQueues>,
+}
+
+impl StreamSocket {
+    pub fn request_stream<T>(&self, stream_id: u16) -> StreamSender<T> {
```

Comment:
How about an enum or generally something more strongly typed for the stream id? Enums don't seem like an issue for packet stuff, in fact they might be an advantage if we would hypothetically want to add or remove streams between versions w/o protocol break (e.g. putting higher frequency reduced tracking on it's own thing or w.e.) so we can actually explicitly handle that and discard packets and not accumulate them.

Reply:
I did use associated constants at some point. but then I removed them. I forgot the exact reason but it wouldn't fit anymore a refactoring I did later, or someone told me that it was a bad design